src/interp/interp.cc

```diff
@@ -260,31 +268,46 @@ void Store::Collect() {
     thread->Mark();
   }
 
-  // TODO: better GC algo.
-  // Loop through all newly marked objects and mark their referents.
-  std::vector<bool> all_marked(object_count, false);
-  bool new_marked;
-  do {
-    new_marked = false;
-    for (size_t i = 0; i < object_count; ++i) {
-      if (!all_marked[i] && marks_[i]) {
-        all_marked[i] = true;
-        objects_.Get(i)->Mark(*this);
-        new_marked = true;
-      }
-    }
-  } while (new_marked);
+  // This vector is often empty since the default maximum
+  // recursion is usually enough to mark all objects.
+  while (!context.unmarked_objects.empty()) {
+    size_t index = context.unmarked_objects.back();
+
+    assert(context.marks[index]);
+    assert(context.call_depth == 0);
+
+    context.unmarked_objects.pop_back();
+    objects_.Get(index)->Mark(*this);
+  }
+
+  assert(context.call_depth == 0);
+
+  gc_context_ = nullptr;
 
   // Delete all unmarked objects.
   for (size_t i = 0; i < object_count; ++i) {
-    if (objects_.IsUsed(i) && !all_marked[i]) {
+    if (objects_.IsUsed(i) && !context.marks[i]) {
       objects_.Delete(i);
     }
   }
 }
 
 void Store::Mark(Ref ref) {
-  marks_[ref.index] = true;
+  size_t index = ref.index;
+
+  if (gc_context_->marks[index])
+    return;
+
+  gc_context_->marks[index] = true;
+
+  if (gc_context_->call_depth >= max_call_depth) {
+    gc_context_->unmarked_objects.push_back(index);
```

Comment:
It looks like `index` is "marked" above on line 301 before being added to "unmarked_objects" .. perhaps this vector should be called something something more accurate.. obviously "marked_objects_that_have_not_yet_been_recursively_traced" won't work.   How about "untraced_objects"?

Reply:
Good idea!